src/main/java/com/axonivy/market/service/impl/ProductServiceImpl.java

```diff
@@ -0,0 +1,251 @@
+package com.axonivy.market.service.impl;
+
+import static org.apache.commons.lang3.StringUtils.EMPTY;
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import org.apache.commons.lang3.StringUtils;
+import org.kohsuke.github.GHCommit;
+import org.kohsuke.github.GHContent;
+import org.kohsuke.github.GHRepository;
+import org.kohsuke.github.GHTag;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageImpl;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.domain.Sort;
+import org.springframework.data.domain.Sort.Order;
+import org.springframework.stereotype.Service;
+import org.springframework.util.CollectionUtils;
+
+import com.axonivy.market.constants.GitHubConstants;
+import com.axonivy.market.entity.GithubRepoMeta;
+import com.axonivy.market.entity.Product;
+import com.axonivy.market.enums.FileType;
+import com.axonivy.market.enums.FilterType;
+import com.axonivy.market.enums.SortOption;
+import com.axonivy.market.factory.ProductFactory;
+import com.axonivy.market.github.model.GitHubFile;
+import com.axonivy.market.github.service.GHAxonIvyMarketRepoService;
+import com.axonivy.market.github.service.GithubService;
+import com.axonivy.market.github.util.GithubUtils;
+import com.axonivy.market.repository.GithubRepoMetaRepository;
+import com.axonivy.market.repository.ProductRepository;
+import com.axonivy.market.service.ProductService;
+
+import lombok.extern.log4j.Log4j2;
+
+@Log4j2
+@Service
+public class ProductServiceImpl implements ProductService {
+
+  private final ProductRepository productRepository;
+  private final GHAxonIvyMarketRepoService axonivyMarketRepoService;
+  private final GithubRepoMetaRepository githubRepoMetaRepository;
+  private final GithubService githubService;
+
+  private GHCommit lastGHCommit;
+  private GithubRepoMeta marketRepoMeta;
+
+  public ProductServiceImpl(ProductRepository productRepository, GHAxonIvyMarketRepoService axonivyMarketRepoService,
+      GithubRepoMetaRepository githubRepoMetaRepository, GithubService githubService) {
+    this.productRepository = productRepository;
+    this.axonivyMarketRepoService = axonivyMarketRepoService;
+    this.githubRepoMetaRepository = githubRepoMetaRepository;
+    this.githubService = githubService;
+  }
+
+  @Override
+  public Page<Product> findProducts(String type, String keyword, Pageable pageable) {
+    final var filterType = FilterType.of(type);
+    if (StringUtils.isNoneBlank(keyword)) {
+      return searchProducts(filterType, keyword, pageable);
+    }
+
+    syncLatestDataFromMarketRepo();
```

Comment:
Why do we only sync data when searching without keywords?

Reply:
the idea is: that "search" should have the fastest response, and search only be fired at the second call from UI - then we are always on DB => resolve performance stuff.
Otherwise, we will fetch data from every sources such as GitHub, DB and maybe later we have more